### DISCUSSION TITLE
`best_match` behavior

### DISCUSSION THREAD

**Discussion by azmeuk** [pallets/werkzeug#2731] [Q&A]
`best_match` have a behavior that feels strange to me: when a preferred *regionalized locale* is not available (say `fr-FR`), I would expect the *unregionalized locale* to be returned if available (`fr`).

```python
# data come from a real life situation I encountered
>>> from werkzeug.datastructures import LanguageAccept
>>> accept = LanguageAccept([('fr-FR', 1), ('en-US', 0.7), ('en', 0.3)])
>>> available = ['gl', 'de', 'fr', 'es', 'nb_NO', 'en']
>>> accept.best_match(available)
'en' # <-- I would have preferred 'fr'
```

I am not sure if this is a bug or a wanted undocumented behavior.
What do you think?

Related tickets #1651 #450

**Comment by davidism** [comment]
It looks like https://github.com/pallets/werkzeug/pull/1507 is the current behavior. From my last comment there, it does seem like `fr` should match since `fr-FR` is the highest priority.

**Reply by davidism** [reply]
Converted to issue https://github.com/pallets/werkzeug/issues/2732

**Comment by davidism** [comment]
`en` is in the accept list and available list, so the exact match is preferred. This behavior won't change, although you could use your own `LanguageAccept` subclass to do what you want.

**Comment by Gidgidonihah** [comment]
The issue to which it was converted is closed, so I'll add the comment here.

I would argue that `best_match` is a `lookup` function, and not a `filtering` function. [RFC4676 section 3.4](https://www.rfc-editor.org/rfc/rfc4647#section-3.4) describes a lookup and indicates that `fr` should match over `en`.

It seems that the way `best_match` logic is implemented is instead [filtering from RFC4676 section 3.3](https://www.rfc-editor.org/rfc/rfc4647#section-3.3) but with only one result returned, rather than all matches.
